### ISSUE TITLE
Getting UNKNOWN_ENUM: 0xc0000466

### ISSUE THREAD

**Issue by FromZeus** [jborean93/smbprotocol#44]
### Description

When I'm trying to connect to the server with NTLM I'm getting 0xc0000466 error which is STATUS_SERVER_UNAVAILABLE. But the server is available and I'm successfully connecting through smbclient for example:
```
Starting GENSEC submechanism ntlmssp
Got challenge flags:
Got NTLMSSP neg_flags=0x60898215
  NTLMSSP_NEGOTIATE_UNICODE
  NTLMSSP_REQUEST_TARGET
  NTLMSSP_NEGOTIATE_SIGN
  NTLMSSP_NEGOTIATE_NTLM
  NTLMSSP_NEGOTIATE_ALWAYS_SIGN
  NTLMSSP_TARGET_TYPE_DOMAIN
  NTLMSSP_NEGOTIATE_EXTENDED_SESSIONSECURITY
  NTLMSSP_NEGOTIATE_TARGET_INFO
  NTLMSSP_NEGOTIATE_128
  NTLMSSP_NEGOTIATE_KEY_EXCH
NTLMSSP: Set final flags:
Got NTLMSSP neg_flags=0x62088215
  NTLMSSP_NEGOTIATE_UNICODE
  NTLMSSP_REQUEST_TARGET
  NTLMSSP_NEGOTIATE_SIGN
  NTLMSSP_NEGOTIATE_NTLM
  NTLMSSP_NEGOTIATE_ALWAYS_SIGN
  NTLMSSP_NEGOTIATE_EXTENDED_SESSIONSECURITY
  NTLMSSP_NEGOTIATE_VERSION
  NTLMSSP_NEGOTIATE_128
  NTLMSSP_NEGOTIATE_KEY_EXCH
NTLMSSP Sign/Seal - Initialising with flags:
Got NTLMSSP neg_flags=0x62088215
  NTLMSSP_NEGOTIATE_UNICODE
  NTLMSSP_REQUEST_TARGET
  NTLMSSP_NEGOTIATE_SIGN
  NTLMSSP_NEGOTIATE_NTLM
  NTLMSSP_NEGOTIATE_ALWAYS_SIGN
  NTLMSSP_NEGOTIATE_EXTENDED_SESSIONSECURITY
  NTLMSSP_NEGOTIATE_VERSION
  NTLMSSP_NEGOTIATE_128
  NTLMSSP_NEGOTIATE_KEY_EXCH
^[[ANTLMSSP Sign/Seal - Initialising with flags:
Got NTLMSSP neg_flags=0x62088215
  NTLMSSP_NEGOTIATE_UNICODE
  NTLMSSP_REQUEST_TARGET
  NTLMSSP_NEGOTIATE_SIGN
  NTLMSSP_NEGOTIATE_NTLM
  NTLMSSP_NEGOTIATE_ALWAYS_SIGN
  NTLMSSP_NEGOTIATE_EXTENDED_SESSIONSECURITY
  NTLMSSP_NEGOTIATE_VERSION
  NTLMSSP_NEGOTIATE_128
  NTLMSSP_NEGOTIATE_KEY_EXCH
 session setup ok
signed SMB2 message
signed SMB2 message
 tconx ok
```

### Log
```
INFO:smbprotocol.connection:Initialising connection, guid: 3683b9d7-c080-4104-84db-012322e8d87d, require_singing: True, server_name: xxx, port: 445
INFO:smbprotocol.connection:Setting up transport connection
INFO:smbprotocol.connection:Starting negotiation with SMB server
INFO:smbprotocol.connection:Negotiating with SMB2 protocol with highest client dialect of: SMB_3_1_1
INFO:smbprotocol.connection:Sending SMB2 Negotiate message
INFO:smbprotocol.transport:Connecting to DirectTcp socket
INFO:smbprotocol.connection:Receiving SMB2 Negotiate response
INFO:smbprotocol.connection:Negotiated dialect: (785) SMB_3_1_1
INFO:smbprotocol.connection:Connection require signing: True
INFO:smbprotocol.session:Initialising session with username: xxx
INFO:smbprotocol.session:Attempting auth with mech KRB5 (1.2.840.113554.1.2.2)
WARNING:smbprotocol.session:Failed auth for mech KRB5 (1.2.840.113554.1.2.2): Mech Type 1.2.840.113554.1.2.2 is not yet supported
INFO:smbprotocol.session:Attempting auth with mech NTLMSSP (1.3.6.1.4.1.311.2.2.10)
INFO:smbprotocol.session:Setting up NTLM Security Context for user xxx
INFO:smbprotocol.session:NTLM: Generating Negotiate message
INFO:smbprotocol.session:Sending SMB2_SESSION_SETUP request message
INFO:smbprotocol.session:Receiving SMB2_SESSION_SETUP response message
WARNING:smbprotocol.session:Failed auth for mech NTLMSSP (1.3.6.1.4.1.311.2.2.10): Received unexpected status from the server: (3221226598) UNKNOWN_ENUM: 0xc0000466
Traceback (most recent call last):
  File "main.py", line 171, in <module>
    main()
  File "main.py", line 155, in main
    concats.append(process(**el))
  File "main.py", line 40, in process
    df = concat_block(paths, sheets, columns, spec)
  File "main.py", line 53, in concat_block
    p, sheet_name=s, header=spec["header"] if "header" in spec else 0)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/util/_decorators.py", line 296, in wrapper
    return func(*args, **kwargs)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/io/excel/_base.py", line 304, in read_excel
    io = ExcelFile(io, engine=engine)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/io/excel/_base.py", line 867, in __init__
    self._reader = self._engines[engine](self._io)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/io/excel/_xlrd.py", line 22, in __init__
    super().__init__(filepath_or_buffer)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/io/excel/_base.py", line 344, in __init__
    filepath_or_buffer, _, _, _ = get_filepath_or_buffer(filepath_or_buffer)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/io/common.py", line 222, in get_filepath_or_buffer
    filepath_or_buffer, mode=mode or "rb", **(storage_options or {})
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/core.py", line 399, in open
    **kwargs
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/core.py", line 250, in open_files
    expand=expand,
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/core.py", line 570, in get_fs_token_paths
    fs = cls(**options)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/spec.py", line 55, in __call__
    obj = super().__call__(*args, **kwargs)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/implementations/smb.py", line 101, in __init__
    self._connect()
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/implementations/smb.py", line 110, in _connect
    connection_timeout=self.timeout,
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/smbclient/_pool.py", line 113, in register_session
    session.connect()
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/smbprotocol/session.py", line 319, in connect
    "%s" % str(errors))
smbprotocol.exceptions.SMBAuthenticationError: Failed to authenticate with server: {'KRB5 (1.2.840.113554.1.2.2)': 'Mech Type 1.2.840.113554.1.2.2 is not yet supported', 'NTLMSSP (1.3.6.1.4.1.311.2.2.10)': 'Received unexpected status from the server: (3221226598) UNKNOWN_ENUM: 0xc0000466'}
INFO:smbprotocol.connection:Disconnecting transport connection
INFO:smbprotocol.transport:Disconnecting DirectTcp socket
```
**System**: Ubuntu 20.04.1 LTS

**Comment by jborean93** [comment]
I'm unsure why this is failing but there are a few things you can try/look into

* I've recently merged https://github.com/jborean93/smbprotocol/pull/41 which changes the underlying auth handler to something more Negotiate like
    * I doubt it fixes the problem but it's worth trying out and seeing if it's fixed
* I see in your `smbclient` output it has the `NTLMSSP_TARGET_TYPE_DOMAIN` flag meaning you are authenticating with a domain account
    * How have you defined the username, is it in the netlogon form `DOMAIN\username`, upn form `username@DOMAIN.COM` or just the `username`?
    * Does the username contain and non-ascii chars?
* Because it does look like a domain account have you considered trying to get Kerberos auth working, it's a lot more secure than NTLM and provides server authentication

What's the target SMB server, is it a Windows host, what's the OS version? If it's a Windows target I would have a look at the security event logs and see if it contains more info as to why the authentication was rejected. Based on the debug logs you've shared it's failing straight after it send the NTLM Negotiate message and no NTLM Challenge message was received.

**Comment by FromZeus** [comment]
Hey @jborean93, thank you for the tips!

- I've tried to setup the latest version but it didn't fix the problem for me unfortunately.
- Yes, I've been using username@DOMAIN notation and tried with DOMAIN/username also with no non-ascii chars.
- Looks like the server doesn't support it.

I think it's a Windows host but I don't have access to it. Here's the log with the latest version of the library from master branch.

```
INFO:smbprotocol.connection:Initialising connection, guid: 88244a8f-84fd-42e1-ac3b-dd100fb053c5, require_singing: True, server_name: xxx, port: 445
INFO:smbprotocol.connection:Setting up transport connection
INFO:smbprotocol.connection:Starting negotiation with SMB server
INFO:smbprotocol.connection:Negotiating with SMB2 protocol with highest client dialect of: SMB_3_1_1
INFO:smbprotocol.connection:Sending SMB2 Negotiate message
INFO:smbprotocol.transport:Connecting to DirectTcp socket
INFO:smbprotocol.connection:Receiving SMB2 Negotiate response
INFO:smbprotocol.connection:Negotiated dialect: (785) SMB_3_1_1
INFO:smbprotocol.connection:Connection require signing: True
INFO:smbprotocol.session:Initialising session with username: xxx@xxx
INFO:smbprotocol.session:Sending SMB2_SESSION_SETUP request message
INFO:smbprotocol.session:Receiving SMB2_SESSION_SETUP response message
INFO:smbprotocol.session:More processing is required for SMB2_SESSION_SETUP
INFO:smbprotocol.session:Sending SMB2_SESSION_SETUP request message
INFO:smbprotocol.session:Receiving SMB2_SESSION_SETUP response message
Traceback (most recent call last):
  File "main.py", line 159, in <module>
    main()
  File "main.py", line 143, in main
    concats.append(process(**el))
  File "main.py", line 42, in process
    df = concat_block(paths, sheets, columns, spec)
  File "main.py", line 55, in concat_block
    p, sheet_name=s, header=spec["header"] if "header" in spec else 0)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/util/_decorators.py", line 296, in wrapper
    return func(*args, **kwargs)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/io/excel/_base.py", line 304, in read_excel
    io = ExcelFile(io, engine=engine)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/io/excel/_base.py", line 867, in __init__
    self._reader = self._engines[engine](self._io)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/io/excel/_xlrd.py", line 22, in __init__
    super().__init__(filepath_or_buffer)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/io/excel/_base.py", line 344, in __init__
    filepath_or_buffer, _, _, _ = get_filepath_or_buffer(filepath_or_buffer)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/pandas/io/common.py", line 222, in get_filepath_or_buffer
    filepath_or_buffer, mode=mode or "rb", **(storage_options or {})
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/core.py", line 399, in open
    **kwargs
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/core.py", line 250, in open_files
    expand=expand,
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/core.py", line 570, in get_fs_token_paths
    fs = cls(**options)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/spec.py", line 55, in __call__
    obj = super().__call__(*args, **kwargs)
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/implementations/smb.py", line 101, in __init__
    self._connect()
  File "/home/dtrishkin/.pyenv/versions/3.7.8/lib/python3.7/site-packages/fsspec/implementations/smb.py", line 110, in _connect
    connection_timeout=self.timeout,
  File "/tmp/smbprotocol/smbclient/_pool.py", line 119, in register_session
    session.connect()
  File "/tmp/smbprotocol/smbprotocol/session.py", line 278, in connect
    raise exc
  File "/tmp/smbprotocol/smbprotocol/session.py", line 275, in connect
    response = self.connection.receive(request)
  File "/tmp/smbprotocol/smbprotocol/connection.py", line 763, in wrapped
    return func(self, *args, **kwargs)
  File "/tmp/smbprotocol/smbprotocol/connection.py", line 1075, in receive
    raise SMBResponseException(response, status)
smbprotocol.exceptions.SMBResponseException: Received unexpected status from the server: (3221226598) UNKNOWN_ENUM: 0xc0000466
INFO:smbprotocol.connection:Disconnecting transport connection
INFO:smbprotocol.transport:Disconnecting DirectTcp socket
```

**Comment by jborean93** [comment]
Ok the debug logs there indicate it at least produced the NTLM Challenge and then failed with `STATUS_SERVER_UNAVAILABLE` when it received the final NTLM Authenticate message. Unfortunately without more info on the target SMB server it will be difficult to determine why it's responding with that error sorry.

**Comment by jborean93** [comment]
I've just created the [v1.2.0 release](https://github.com/jborean93/smbprotocol/releases/tag/v1.2.0) and while it does not fix your issue it has registered this error as [ServerUnavailable](https://github.com/jborean93/smbprotocol/blob/a7d0f219e48d58b0fb9295a460ff76c698b53413/smbprotocol/exceptions.py#L551-L553) in the exception types.

Ultimately the error is being returned by the server so it would have to be checked to see why it's returning that value. That's not something I can do myself. If you do find out more info or have replicated the error in some way that I can follow please comment. Unfortunately in the meantime there's not much else I can do to help you unfortunately so I'll close this issue for now.

**Comment by jborean93** [comment]
It looks like this was ultimately an issue in pyspnego with how it handled Negotiate authentication. The details of this problem can be found at https://github.com/jborean93/smbprotocol/issues/61 and the fix is included in [pyspnego 0.1.4](https://github.com/jborean93/pyspnego/releases/tag/v0.1.4).
